### PR TITLE
Computation of negative powers of a quaternion

### PR DESCRIPTION
**Pull request by arighnac**
Fixes issue #13688 

This commit now helps to calculate the negative power of a quaternion (p<-1) , which was previously being shown as -1.

**PREVIOUS**

 `>>>q` `=` `Quaternion(1,2,3,4)`
`>>>q.pow(-2)`
`>>>1`

**NOW**

`>>>q = Quaternion(1,2,3,4)`
 `>>>q.pow(-2)`
 `>>>(-7/225) + (-1/225)*i + (-1/150)*j + (-2/225)*k`

@gxyd , @jksuom , @ylemkimon  please review.

### CODE REVIEW

File: sympy/algebras/quaternion.py [hunk 1]
```diff
@@ -328,6 +328,15 @@ def pow(self, p):
...
             q = q * q
 
+        if p < -1:
+            r = abs(p)
+            while r > 0:
+                if r & 1:
+                    res = q.inverse() * res
+
+                r = r >> 1
+                q.inverse() = q.inverse() * q.inverse()
```
Comment (gxyd) [inline sympy/algebras/quaternion.py]: I don't think this is a valid operation. Assigning to a method is not valid (atleast not in python).

**Review by gxyd — COMMENTED** [review]
Can you make sure that your PR only contains valid commits? Ylemkimon suggested a possible reason for this to you, I think that is the thing you are doing wrong.

**Comment by gxyd** [comment]
Also the 'Fixes #NNNN' doesn't work with 'Fixes issue #NNNN', please make sure to use 'Fixes #NNNN' or 'Resolves #NNNN' in your PR description.

**Comment by ArighnaIITG** [comment]
I have done literally everything. But nothing seems to remove those commits which are appearing above.
I have even done git cherry-pick ".....", but even that fails to work.
git stash and git rebase -i~HEAD-n also fails to work.
Infact the above commits had not been done by me.

@asmeurer , @jksuom , @ylemkimon , @gxyd , @smichr , please help.

**Review by gxyd — COMMENTED** [review]
Also please add tests cases for this.

File: sympy/algebras/quaternion.py [hunk 2]
```diff
@@ -321,13 +321,25 @@ def pow(self, p):
...
         res = 1
+
         while p > 0:
             if p & 1:
                 res = q * res
 
             p = p >> 1
             q = q * q
 
+        if p < -1:
```
Comment (gxyd) [inline sympy/algebras/quaternion.py]: Will `q_inv = q.inverse()` and then returning `q_inv.power(-p)` work?

Comment (jksuom) [reply sympy/algebras/quaternion.py]: Maybe this could be before the loop above:

    if p < 0:
        q, p = q.inverse(), -p

Comment (gxyd) [reply sympy/algebras/quaternion.py]: Yes, ofcourse. That will work better.

File: sympy/algebras/tests/test_quaternion.py [hunk 3]
```diff
@@ -56,6 +56,7 @@ def test_quaternion_functions():
     assert q.normalize() == Quaternion(x, y, z, w) / sqrt(w**2 + x**2 + y**2 + z**2)
     assert q.inverse() == Quaternion(x, -y, -z, -w) / (w**2 + x**2 + y**2 + z**2)
     assert q.pow(2) == Quaternion(-w**2 + x**2 - y**2 - z**2, 2*x*y, 2*x*z, 2*w*x)
+    assert q1.pow(-2) == Quaternion(-(7/225), -(1/225), -(1/150), -(2/225))
```
Comment (gxyd) [inline sympy/algebras/tests/test_quaternion.py]: use `-S(7)/225` instead. See:
```python
>>> type(7 / 25)
flaot
>>> type(S(7) / 25)
sympy.core.numbers.Rational
```

**Review by gxyd — COMMENTED** [review]
Currently:
```python
In [1]: a = Quaternion(1, 2, 1, 2)

In [2]: a**(S(-1)/2)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-e58ad0a9acfc> in <module>()
----> 1 a**(S(-1)/2)

/home/gxyd/dev/sympy/sympy/algebras/quaternion.py in __pow__(self, p)
    165 
    166     def __pow__(self, p):
--> 167         return self.pow(p)
    168 
    169     def __neg__(self):

/home/gxyd/dev/sympy/sympy/algebras/quaternion.py in pow(self, p)
    327 
    328         while p > 0:
--> 329             if p & 1:
    330                 res = q * res
    331 

TypeError: unsupported operand type(s) for &: 'Half' and 'int'

```

I would expect a better error message.

**Comment by ArighnaIITG** [comment]
@gxyd ,
Can you tell more about the "better error message" which you were talking above.?
Like, where to make the changes?

**Comment by gxyd** [comment]
Seeing on the wikipedia page https://en.wikipedia.org/wiki/Quaternion#Exponential,_logarithm,_and_power . Currently it doesn't deal well with powers of non-integers quaternion.

File: sympy/algebras/tests/test_quaternion.py (hunk 3 above)
Comment (ArighnaIITG) [reply sympy/algebras/tests/test_quaternion.py]: @gxyd , @jksuom ,
Could you please tell as to what more to do here?
I could not understand where to make the code change for a new error message.
Please help.

**Comment by ylemkimon** [comment]
You can check for type of p before computation and return `NotImplemented`, or implement non-integer power.

**Comment by ArighnaIITG** [comment]
@ylemkimon , I have added NotImplemented. 
Please review.

File: sympy/algebras/quaternion.py [hunk 4]
```diff
@@ -321,6 +321,13 @@ def pow(self, p):
         if p == -1:
             return q.inverse()
         res = 1
+
+        if p < 0:
+            q, p = q.inverse(), -p
+
+        if (type(p) != int):
```
Comment (ylemkimon) [inline sympy/algebras/quaternion.py]: Use `isinstance` to check for its type. Also, check for other integer types, like SymPy integer, `Integer`. Consider using `SYMPY_INTS` from `sympy.core.compatibility`.

**Review by ylemkimon — CHANGES_REQUESTED** [review]

**Comment by ArighnaIITG** [comment]
@ylemkimon ,
I have added an isinstance check to check the type. Please review.

File: sympy/algebras/quaternion.py (hunk 4 above)
Comment (ArighnaIITG) [reply sympy/algebras/quaternion.py]: Thanks. It has been implemented. Please review.

File: sympy/algebras/quaternion.py [hunk 5]
```diff
@@ -321,6 +323,13 @@ def pow(self, p):
         if p == -1:
             return q.inverse()
         res = 1
+
+        if p < 0:
+            q, p = q.inverse(), -p
+
+        if not (isinstance(p, (int, Integer, SYMPY_INTS))):
```
Comment (ylemkimon) [inline sympy/algebras/quaternion.py]: `SYMPY_INTS` already includes `int`.

**Review by ylemkimon — CHANGES_REQUESTED** [review]
Otherwise LGTM

File: sympy/algebras/quaternion.py (hunk 5 above)
Comment (ArighnaIITG) [reply sympy/algebras/quaternion.py]: So, will I remove int? @ylemkimon

**Review by ylemkimon — APPROVED** [review]

**Comment by ArighnaIITG** [comment]
@smichr , @gxyd 
This has been approved by @ylemkimon.
I think this can be merged now.
Thank you.

**Comment by ylemkimon** [comment]
Thank you! It's in.

**Pull request merged**